### ISSUE TITLE
pydantic v2 memory usage with hundreds of complicated models

### ISSUE THREAD

**Issue by MarshalX** [pydantic/pydantic#9982]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I tried to simplify the large number of complicated models as much as possible. 412 Pydantic models share the same configuration, using strict mode and population by name. The models extensively employ Union and Tagged Union types. They are loaded in the main module and then rebuilt in a separate module. The memory usage increases by over 100MB for all models upon startup.

I want to clarify that this isn’t a formal bug report, but I am curious whether this memory usage is expected and if there are any plans for optimization in the future. Thank you!

### Example Code

I prepared the repository to reproduce it: https://github.com/MarshalX/pydantic-memory-usage

### Python, Pydantic & OS Version

```text
             pydantic version: 2.9.0
        pydantic-core version: 2.23.2
          pydantic-core build: profile=release pgo=false
                 install path: blabla/projects/pydantic-memory-usage/venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.6 (default, Mar 29 2024, 10:51:09)  [Clang 15.0.0 (clang-1500.3.9.4)]
                     platform: macOS-14.6.1-arm64-arm-64bit
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@MarshalX,

Thanks for reporting this issue. We're currently investigating reusing common types + schemas to improve performance and memory usage. We'll be sure to use this as a reference / benchmark!

**Comment by sydney-runkle** [comment]
Hi @MarshalX,

Could you test this against the branch associated with https://github.com/pydantic/pydantic/pull/10064?

**Comment by MarshalX** [comment]
Hi @sydney-runkle,

Done. I've updated repo's README with new results

**Comment by sydney-runkle** [comment]
Hey @MarshalX,

What about with the code associated with this PR? https://github.com/pydantic/pydantic/pull/10113

**Comment by MarshalX** [comment]
Hey @sydney-runkle,

Done. I do not see any difference. Do I do something wrong? I use `pip install git+https://github.com/pydantic/pydantic.git@namespace-change` and nothing additional

**Comment by sydney-runkle** [comment]
Ah I think it's because you're importing the models from within a function rather than at the module level:

```py
@profile
def __rebuild_all_models() -> None:
    # load models to the scope
    import models  # noqa
    from models.unknown_type import UnknownType, UnknownInputType  # noqa

    for __model in __get_models_to_rebuild_set():
        __model.model_rebuild()
```

Because of this, I don't think the namespace change will have an effect. What if you try running that code at the module level? Thanks for helping us with this benchmarking!

**Comment by sydney-runkle** [comment]
Hey @MarshalX, were you able to try the above suggestion?

**Comment by MarshalX** [comment]
Hey @sydney-runkle , I have tried your suggestion but it fails with `NameError: name 'models' is not defined` when I try to move it to the module level. Probably that's why I ended up with moving it to the function ~ a year ago.

**Comment by sydney-runkle** [comment]
Hmm, seems like a bug. Is there anyw ay you could give us a minimal snippet to reproduce that issue?

**Comment by MarshalX** [comment]
@sydney-runkle started to prepare a minimal snippet to reproduce, got it working, then updated to 2.9.0, and can't reproduce anymore! Downgraded to 2.8.2 and can reproduce. So somehow was fixed already :)

**Comment by MarshalX** [comment]
@sydney-runkle I've added the results with v2.9.0. But I did not use `experimental_fast_build` because I was confused by some cherry-picking and couldn't find the proper env var how to enable it. Also, I did not move model rebuilding to the module scope. It is still relevant?

**Comment by MarshalX** [comment]
Yes you are correct. No difference :(

**Comment by sydney-runkle** [comment]
Hmph :( ok. Will continue to chip away here in the coming months!

**Comment by MarshalX** [comment]
Hi @sydney-runkle! Do we have any updates here?

**Comment by atti92** [comment]
I was thinking if pydantic could do something about FastAPI's TypeAdapter memory usage problem. 

It's causing us a lot of headache to hack around the fact FastAPI is not caching TypeAdapters, and seems to make duplications of them leading to very high memory usages. 

Is it technically possible for you to implement that inside pydantic to reuse the same memory if TypeAdapters share the same characteristics?

**Comment by sydney-runkle** [comment]
We're working on memory usage for our v2.11 release which involves caching core schemas / pydantic core data structures when possible. We'll provide updates here when we have PRs closer to merging!

**Comment by atti92** [comment]
It sounds really nice. 

Our use-case includes a really big model, with a lot of deep submodels, some of which we also use separately in our requests/responses in FastAPI. And those models are also included in our high level data models that use Generic typing, like:
```
class DTO(Generic[T]):
    data: T
    meta: MetaType
```
and we have like 4 of these classes which use `T` in different structures, and also in list[T], etc...

And T can be a couple of the models we use all of which are pretty complex, and one of them includes all the other in some level.

Sadly FastAPI only caches by the Top level model, and creates TypeAdapters for all fields in them unconditionally.
All of this made our server really memory hungry, by doing a monkeypatch on FastAPI we could reduce memory usage by 550MB (cache all typeadapters by type and metadata)

I hope if you can do it on low level it would be a really huge memory save for us.

**Comment by sydney-runkle** [comment]
@atti92,

What sort of memory usage are you seeing on v2.10.1?

**Comment by atti92** [comment]
All RSS values measured without load, after service startup:
- Pydantic: `2.10.1`: `780 MB`
- Pydantic: `2.10.1` + FastAPI TypeAdapter cache hack: `276 MB`
- Pydantic: `2.9.2`: `806 MB`
- Pydantic: `2.9.2` + FastAPI TypeAdapter cache hack: `280 MB`

I did each test twice and didn't see any variance in the values.

**Comment by sydney-runkle** [comment]
Yes, we're actively working on perf for v2.11. Will update here with specifics once we pick this issue back up.

**Comment by Viicos** [comment]
2.11 (final released planned by the end of March) brings significant performance improvements related to constructing models, in particular core schema building [^1].

An alpha release (2.11.0a1) was released today and can be installed, so that you can try it out. Do not hesitate to provide feedback if this improved (or did not) performance on your side.

Note that memory consumption haven’t been really optimized yet in this alpha release. We have plans to optimize it before the final release.

[^1]: Pydantic defines a core schema whenever a Model is defined, read more [here](https://docs.pydantic.dev/latest/internals/architecture/)

**Comment by Viicos** [comment]
> Hello! I don’t see any changes in RAM usage based on the benchmark provided by MarshalX in version 2.11.0a1.

Yes, this is expected for the first alpha. I have a marshalx benchmark locally, I'll take a look.

**Comment by Viicos** [comment]
I would suggest running your test suite and make sure everything is working, and testing in a dev/staging environment first. But if memory usage is the only concern, then it's not worth the effort, as we haven't tackled it yet in this alpha release.

**Comment by sydney-runkle** [comment]
Great to see. We can continue to strive for memory improvements, though this was an important first step.

**Comment by MarshalX** [comment]
i have updated my test repo with the results of 2.11.2. this is impressive! uses half of memory and x2 faster to import. well, done pydantic team! thank you! Should we close it or any improvements are coming soon?

results: https://github.com/MarshalX/pydantic-memory-usage/commit/f3f2d3600ba1148d4abbf8961a128f028c8fdb9a

**Comment by Viicos** [comment]
Thanks for producing the repro. I think we can close for now. I believe more improvements could be made in the future (probably on memory consumption by reusing more validators), but this would require significant refactors in the schema building process.
